.builder/jdk.py

```diff
@@ -0,0 +1,120 @@
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import glob
+import os
+from pathlib import Path
+import tarfile
+import zipfile
+from urllib.request import urlretrieve
+
+import Builder
+util = Builder.Util
+
+URLs = {
+    'linux-armv6': 'https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u232-b09/OpenJDK8U-jdk_arm_linux_hotspot_8u232b09.tar.gz',
+    'linux-armv7': 'https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u232-b09/OpenJDK8U-jdk_arm_linux_hotspot_8u232b09.tar.gz',
+    'linux-arm64': 'https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u242-b08/OpenJDK8U-jdk_aarch64_linux_hotspot_jdk8u242-b08.tar.gz',
+    'linux-x64': 'https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u242-b08/OpenJDK8U-jdk_x64_linux_hotspot_8u242b08.tar.gz',
+    'windows-x64': 'https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u242-b08/OpenJDK8U-jdk_x64_windows_hotspot_8u242b08.zip',
+    'windows-x86': 'https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u242-b08/OpenJDK8U-jdk_x86-32_windows_hotspot_8u242b08.zip',
+    'macos-x64': 'https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u242-b08/OpenJDK8U-jdk_x64_mac_hotspot_8u242b08.tar.gz',
+}
+
+
+class JDK8(Builder.Import):
+    def __init__(self, **kwargs):
+        super().__init__(
+            config={},
+            **kwargs)
+        self.path = None
+        self.installed = False
+
+    def resolved(self):
+        return True
+
+    def install(self, env):
+        if self.installed:
+            return
+
+        sh = env.shell
+
+        # If this is a local build, check the local machine
+        if not env.toolchain.cross_compile:
+            javac_path = util.where('javac')
+            if javac_path:
+                javac_path = javac_path.replace(
+                    os.pathsep + 'bin' + os.pathsep + '.+$', '')
+            prefixes = [javac_path, os.environ.get('JAVA_HOME', None)]
+            required_files = [
+                ['include/jni.h'],
+                ['lib/**/libjvm.so', '**/lib/**/libjvm.so'],
+            ]
+            found = 0
+            for prefix in prefixes:
+                if not prefix:
+                    continue
+                for paths in required_files:
+                    for path in paths:
+                        full_path = os.path.join(prefix, path)
+                        if glob.glob(full_path, recursive=True):
+                            found += 1
+                            break
+
+                if found >= len(required_files):
+                    print('Found existing JDK8 at {}'.format(prefix))
+                    env.variables['java_home'] = self.path = prefix
+                    self.installed = True
+                    return
+
+        target = '{}-{}'.format(env.spec.target, env.spec.arch)
```

Comment:
trivial: You meant to fall through here if we looked for, but didn't find, an existing local installation? In any case, add a comment where the code resumes explaining the circumstances

`# Install JDK. We didn't find (or we're not using) an existing installation.`

Reply:
There's a return. Basically, there's no way we CAN install the JDK, so I've made it say "please install JDK8 or greater and set JAVA_HOME"